temporal-test-server/src/main/java/io/temporal/internal/testservice/TestWorkflowService.java

```diff
@@ -798,16 +817,92 @@ public void respondNexusTaskFailed(
+        throw Status.INVALID_ARGUMENT
+            .withDescription("Nexus handler error not set on RespondNexusTaskFailedRequest")
+            .asRuntimeException();
+      }
+      NexusOperationRef ref = NexusOperationRef.fromBytes(request.getTaskToken());
+      TestWorkflowMutableState mutableState = getMutableState(ref.getExecutionId());
+      Failure failure = nexusFailureToApplicationFailure(request.getError().getFailure());
+      mutableState.failNexusOperation(ref, failure);
       responseObserver.onNext(RespondNexusTaskFailedResponse.getDefaultInstance());
       responseObserver.onCompleted();
     } catch (StatusRuntimeException e) {
       handleStatusRuntimeException(e, responseObserver);
     }
   }
 
+  public void completeNexusOperation(ByteString token, HistoryEvent completionEvent) {
+    NexusOperationRef ref = NexusOperationRef.fromBytes(token);
+    TestWorkflowMutableState target = getMutableState(ref.getExecutionId());
+
+    switch (completionEvent.getEventType()) {
+      case EVENT_TYPE_WORKFLOW_EXECUTION_COMPLETED:
+        Payloads result =
+            completionEvent.getWorkflowExecutionCompletedEventAttributes().getResult();
+        // All of our SDKs support returning a single value from workflows, we can safely ignore the
+        // rest of the payloads. Additionally, even if a workflow could return more than a single
+        // value,
+        // Nexus does not support it.
+        Payload p =
+            (result.getPayloadsCount() > 0) ? result.getPayloads(0) : Payload.getDefaultInstance();
+        target.completeNexusOperation(ref, p);
+        break;
+      case EVENT_TYPE_WORKFLOW_EXECUTION_FAILED:
+        target.failNexusOperation(
+            ref, completionEvent.getWorkflowExecutionFailedEventAttributes().getFailure());
+        break;
+      case EVENT_TYPE_WORKFLOW_EXECUTION_CANCELED:
+        target.cancelNexusOperation(ref);
+        break;
+      case EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED:
+        Failure terminated =
+            Failure.newBuilder()
+                .setMessage("operation terminated")
+                .setApplicationFailureInfo(
+                    ApplicationFailureInfo.newBuilder().setNonRetryable(true))
+                .build();
+        target.failNexusOperation(ref, terminated);
+        break;
+      case EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT:
+        Failure timedOut =
+            Failure.newBuilder()
+                .setMessage("operation exceeded internal timeout")
+                .setApplicationFailureInfo(
+                    ApplicationFailureInfo.newBuilder().setNonRetryable(true))
+                .build();
+        target.failNexusOperation(ref, timedOut);
+        break;
+      default:
+        throw Status.INTERNAL
+            .withDescription("invalid workflow execution status: " + completionEvent.getEventType())
+            .asRuntimeException();
+    }
+  }
+
+  private static Failure nexusFailureToApplicationFailure(
```

Comment:
Make sure that all failures recorded on events are nexus operation failures.
Sometimes you'll want application failures embedded in those but there should always be a nexus operation failure at the top level:

https://github.com/temporalio/temporal/blob/b383ffffcbbeacdfce2fe021c30f093bab64b5d9/components/nexusoperations/executors.go#L669